### PR TITLE
raftstore: calculate group id for every peer when it is syn-recover status

### LINKED ISSUE #15784: [dr-autosync] hang in sync_recover when change from majority to dr-auto-sync
### What did you do?

1. Config one cluster to dr-auto-sync mode with 3 voters in primary dc and 2 follower + 1 learner in backup dc;
2. Run following steps:
```
	//step 1: down 2 voters in primary dc
	opt.DrClusterInfo.DownNodesInDc(ctxCore, "primary", "tikv", 2)
	c := opt.DrClusterInfo
	ctx := ctxCore
	oltppkg.WaitFor(ctx, "after step 1: down primary 2 tikv, workload running", 1*time.Minute)

	//step 2: change cluster to majority
	output, err := oltppkg.BasicClusterExecPdCtl(c.Cluster, "config set replication-mode majority")
	Expect(err).Should(BeNil())
	Expect(strings.Contains(strings.ToLower(output), "succ")).Should(BeTrue())
	fmt.Println(output)
	oltppkg.WaitFor(ctx, "after step2:change to majority,  workload running", 1*time.Minute)

	//step 3: update placement rule to change from voter to learner for the two down tikv
	c.PdURL = Try(c.Cluster.ServiceURL(resource.PDAddr)).(*url.URL)
	err = drutils.SetPlacementRule(ctx, c.Cluster, c.TiupNode,
		drconfig.PlacementRuleDowngrade)
	Expect(err).Should(BeNil())
	oltppkg.WaitFor(ctx, "after step3: update placement rule, change voter to learner,  workload running", 1*time.Minute)

	//step 4 : clean chaos
	c.CleanChaos()
	oltppkg.WaitFor(ctx, "step 4: clean chaos, waiting for workload running", 1*time.Minute)

	//step 5: set back to dr-auto-sync
	output, err = oltppkg.BasicClusterExecPdCtl(c.Cluster, "config set replication-mode dr-auto-sync")
	Expect(err).Should(BeNil())
	Expect(strings.Contains(strings.ToLower(output), "succ")).Should(BeTrue())
	fmt.Println(output)
	oltppkg.WaitFor(ctx, "step5: set back to dr-autosync workload running", 1*time.Minute)

	//step 6: set back to dr-auto-sync placement rule
	err = drutils.SetPlacementRule(ctx, c.Cluster, c.TiupNode,
		drconfig.PlacementRuleConfigPrimary)
	Expect(err).Should(BeNil())
	drutils.WaitForReplicateStatus(ctx, c.Cluster, 10*time.Minute, 1*time.Minute, "sync")
	oltppkg.WaitFor(gctx, "step 6: set back to dr-auto-sync placement rule, waiting for workload running", 5*time.Minute)
```

### What did you expect to see?
In step 6,  sync_recover can ack to sync after some time (according to the data (10G), should with 10min ).

### What did you see instead?
Hang in sync_recover.

### What version of PD are you using (`pd-server -V`)?
v6.5.5

### PR DESCRIPTION
**Pull request by TonsnakeLin**
### What is changed and how it works?

Issue Number: Close #15784

What's Changed:
In old version, it doesn't re-calculate group-id for the peers created at marjority replication status even it becomes sync-reocver of dr-auto-sync, and the region is always at sync-recover status even the raft-log replication has catched up.
This pr will re-calculate group-id when the current status is sync-reocver, and the region will become sync status when the raft-log replication catches up.

```commit-message
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
       
       //step 1: down 2 voters in primary dc
	opt.DrClusterInfo.DownNodesInDc(ctxCore, "primary", "tikv", 2)
	c := opt.DrClusterInfo
	ctx := ctxCore
	oltppkg.WaitFor(ctx, "after step 1: down primary 2 tikv, workload running", 1*time.Minute)

	//step 2: change cluster to majority
	output, err := oltppkg.BasicClusterExecPdCtl(c.Cluster, "config set replication-mode majority")
	Expect(err).Should(BeNil())
	Expect(strings.Contains(strings.ToLower(output), "succ")).Should(BeTrue())
	fmt.Println(output)
	oltppkg.WaitFor(ctx, "after step2:change to majority,  workload running", 1*time.Minute)

	//step 3: update placement rule to change from voter to learner for the two down tikv
	c.PdURL = Try(c.Cluster.ServiceURL(resource.PDAddr)).(*url.URL)
	err = drutils.SetPlacementRule(ctx, c.Cluster, c.TiupNode,
		drconfig.PlacementRuleDowngrade)
	Expect(err).Should(BeNil())
	oltppkg.WaitFor(ctx, "after step3: update placement rule, change voter to learner,  workload running", 1*time.Minute)

	//step 4 : clean chaos
	c.CleanChaos()
	oltppkg.WaitFor(ctx, "step 4: clean chaos, waiting for workload running", 1*time.Minute)

	//step 5: set back to dr-auto-sync
	output, err = oltppkg.BasicClusterExecPdCtl(c.Cluster, "config set replication-mode dr-auto-sync")
	Expect(err).Should(BeNil())
	Expect(strings.Contains(strings.ToLower(output), "succ")).Should(BeTrue())
	fmt.Println(output)
	oltppkg.WaitFor(ctx, "step5: set back to dr-autosync workload running", 1*time.Minute)

	//step 6: set back to dr-auto-sync placement rule
	err = drutils.SetPlacementRule(ctx, c.Cluster, c.TiupNode,
		drconfig.PlacementRuleConfigPrimary)
	Expect(err).Should(BeNil())
	drutils.WaitForReplicateStatus(ctx, c.Cluster, 10*time.Minute, 1*time.Minute, "sync")
	oltppkg.WaitFor(gctx, "step 6: set back to dr-auto-sync placement rule, waiting for workload running", 5*time.Minute)
       

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

**Review by Connor1996 — COMMENTED** [review]
Please add a test to cover it

File: tests/integrations/raftstore/test_replication_mode.rs [hunk 1]
```diff
@@ -456,6 +468,82 @@ fn test_migrate_replication_mode() {
...
+    assert_eq!(state.state_id, 1);
+    assert_eq!(state.state, RegionReplicationState::IntegrityOverLabel);
+
+    // 2. swith to marjority mode.
+    cluster
+        .pd_client
+        .switch_replication_mode(DrAutoSyncState::Async, vec![]);
+    thread::sleep(Duration::from_millis(150));
+
+    // 3. spilt the region and make a new region, the regions status must be
```
Comment (nolouch) [inline tests/integrations/raftstore/test_replication_mode.rs:510]: current DR will forbid the split during the Async stage. Is there any other situation causing this problem?

Comment (TonsnakeLin) [reply tests/integrations/raftstore/test_replication_mode.rs:510]: um...? it seems to split successfully during the async stage.
I printed the detail of the region_m and region_k.
region_m: id: 1 start_key: 6D31 region_epoch { conf_ver: 1 version: 2 } peers { id: 1 store_id: 1 } peers { id: 2 store_id: 2 } peers { id: 3 store_id: 3 }, 
region_k: id: 1000 end_key: 6D31 region_epoch { conf_ver: 1 version: 2 } peers { id: 1001 store_id: 1 } peers { id: 1002 store_id: 2 } peers { id: 1003 store_id: 3 }

File: tests/integrations/raftstore/test_replication_mode.rs [hunk 2]
```diff
@@ -456,6 +468,82 @@ fn test_migrate_replication_mode() {
...
+    assert_eq!(rx.recv_timeout(Duration::from_millis(100)).is_ok(), true);
+    must_get_equal(&cluster.get_engine(1), b"k2", b"v2");
+    let state = cluster.pd_client.region_replication_status(region.get_id());
+    assert_eq!(state.state_id, 1);
+    assert_eq!(state.state, RegionReplicationState::IntegrityOverLabel);
+
+    // 2. swith to marjority mode.
+    cluster
+        .pd_client
+        .switch_replication_mode(DrAutoSyncState::Async, vec![]);
```
Comment (Connor1996) [inline tests/integrations/raftstore/test_replication_mode.rs:507]: let's make it to marjority, not dr auto sync with async state

File: tests/integrations/raftstore/test_replication_mode.rs (hunk 1 above)
Comment (nolouch) [reply tests/integrations/raftstore/test_replication_mode.rs:510]: split will be forbidden on the pd side, so your test can split with mock pd.

File: tests/integrations/raftstore/test_replication_mode.rs (hunk 2 above)
Comment (Connor1996) [reply tests/integrations/raftstore/test_replication_mode.rs:507]: why can't pd-mock?

File: components/test_pd_client/src/pd.rs [hunk 3]
```diff
@@ -1446,6 +1446,22 @@ impl TestPdClient {
...
 
+    pub fn switch_to_majority_mode(&self) {
+        let mut cluster = self.cluster.wl();
+        let status = cluster.replication_status.as_mut().unwrap();
+        status.set_mode(ReplicationMode::Majority);
+        let mut dr = status.mut_dr_auto_sync();
+        dr.state_id += 1;
+    }
+
+    pub fn switch_to_drautosync_mode(&self) {
```
Comment (Connor1996) [inline components/test_pd_client/src/pd.rs:1457]: you can just make `switch_replication_mode` accept `None`, meaning `Majority`

**Review by nolouch — APPROVED** [review]
lgtm

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**